packages/codemirror/src/editor.ts

```diff
@@ -1410,9 +1410,14 @@ namespace Private {
       'CodeMirror-linenumbers': 'lineNumbers',
       'CodeMirror-foldgutter': 'codeFolding'
     };
-    return Object.keys(classToSwitch).filter(
-      gutter => config[classToSwitch[gutter]]
+
+    // do not remove custom gutters
+    const customGutters = (config.gutters || []).filter(
+      gutter => !classToSwitch[gutter]
     );
+    return Object.keys(classToSwitch)
+      .filter(gutter => config[classToSwitch[gutter]])
+      .concat(customGutters);
```

Comment:
I don't think this respects the order of the gutters as defined in `customGutters`. If one of `CodeMirror-linenumbers` or `CodeMirror-foldgutter` is defined in `config.gutters`, they should be returned in the same order. Otherwise, linenumbers (and probably also folding?) should be added as right-most (last).

Reply:
! I had totally overlooked that. I coded up a solution that naively reflected an order maintaining approach, but I'm not sure it's right behaviour. 

What i'm concerned about is that an extension will specify a gutter order (say ['custom1', 'CodeMirror-linenumbers', 'custom2']), then the user toggles line numbers (yielding ['custom1', 'custom2']), and then toggles again (finally ['custom1', 'custom2', 'CodeMirror-linenumbers']). Then it seems like the extension will have to always be watching the configuration state for changes, which seems kinda fragile. Instead maybe it should be the case that the gutters always occur [PREDEFINED, CUSTOM] (or the reverse), which extensions can then swap via CSS. This solution doesn't allow for interleaving, but I'm not really sure what to do about that. I'm open to any ideas?

Reply:
> What i'm concerned about is that an extension will specify a gutter order (say ['custom1', 'CodeMirror-linenumbers', 'custom2']), then the user toggles line numbers (yielding ['custom1', 'custom2']), and then toggles again (finally ['custom1', 'custom2', 'CodeMirror-linenumbers']).

I'm not sure why this would happen? Does the toggling affect the items in `config.gutters`, or only the booleans in `config.lineNumbers` / `config.codeFolding` ? If the latter, then everything is ok.

Proposed resolution:
- Take `config.gutter` and *append* the keys from `classToSwitch` (this should ensure that they default to *right-most*, as is expected for linenos). The relative order of linenos vs folding should probably be looked at (I see vscode puts their folding to the right of the linenos).
- From this merge, filter the keys from `classToSwitch` according to the mapped config bools. This should be the return value.

This ensures that:
- `config.gutter` is left as is

Reply:
> Does the toggling affect the items in config.gutters, or only the booleans in config.lineNumbers / config.codeFolding ? 

I'm pretty sure it does affect the items in config.gutters? (see particularly the lineNumbers branch of setOption) This functionality is the reason why just inserting custom gutters on the _config doesn't work. 

That said, I'm don't have strong opinions about the way that this functionality should work (other than just not blocking custom gutters). So here is my solution for the above outlined functionality:

```ts
function getActiveGutters(config: CodeMirrorEditor.IConfig): string[] {
    // The order of the classes will be the gutters order
    let classToSwitch: { [val: string]: keyof CodeMirrorEditor.IConfig } = {
      'CodeMirror-linenumbers': 'lineNumbers',
      'CodeMirror-foldgutter': 'codeFolding'
    };
    const configGutters = config.gutters || [];
    const predefinedGuttersToAdd = Object.keys(classToSwitch).filter(gutter => {
      return !configGutters.includes(gutter) && config[classToSwitch[gutter]];
    });

    return configGutters
      .filter(gutter => {
        // if the gutter is custom, then keep it
        const customGutters = !classToSwitch[gutter];
        // keep or remove gutters already specified
        const configSpecified = config[classToSwitch[gutter]];
        return customGutters || configSpecified;
      })
      .concat(predefinedGuttersToAdd);
  }

```
(I have some tests written for this as well, but i'm not entirely sure where to put them? I'm very new to the codebase)

Reply:
That code looks good on a first inspection 👍  Tests would probably best go here: https://github.com/jupyterlab/jupyterlab/blob/master/tests/test-codemirror/src/editor.spec.ts

Reply:
> I'm pretty sure it does affect the items in config.gutters? (see particularly the lineNumbers branch of setOption) This functionality is the reason why just inserting custom gutters on the _config doesn't work.

I agree that the logic here seems overly convoluted, pinging back and forth between the two `setOption` functions. If you see a good way to clean this up, that would be very welcome, or if you just want to leave some comment explaining the current known issues that would also be a clear step up from the current state 😅